pkg/metrics/exporters/stackdriver/stackdriver.go

```diff
@@ -31,26 +41,33 @@ func Start(ctx context.Context) error {
 		return err
 	}
 
-	exporter, err := stackdriver.NewExporter(stackdriver.Options{
-		MetricPrefix:      metricPrefix,
-		MonitoredResource: monitoredresource.Autodetect(),
-	})
+	e, err := stackdriver.New(stackdriver.WithMetricDescriptorTypeFormatter(func(desc metricdata.Metrics) string {
+		return fmt.Sprintf("%s/%s", metricPrefix, desc.Name)
+	}))
 	if err != nil {
 		if *ignoreMissingCreds {
 			log.Error(err, "Error initializing stackdriver exporter, not exporting stackdriver metrics")
 			return nil
 		}
 		return err
 	}
-
-	if err := exporter.StartMetricsExporter(); err != nil {
-		if *ignoreMissingCreds {
-			log.Error(err, "Error starting stackdriver exporter, not exporting stackdriver metrics")
-			return nil
-		}
+	res, err := resource.New(ctx,
+		resource.WithDetectors(gcp.NewDetector(), ec2.NewResourceDetector()),
+		resource.WithTelemetrySDK(),
+		resource.WithFromEnv(),
+	)
+	if err != nil {
 		return err
 	}
-	defer exporter.StopMetricsExporter()
+	reader := metric.NewPeriodicReader(e, metric.WithInterval(*metricInterval))
+	meterProvider := metric.NewMeterProvider(
```

Comment:
Checking on the metrics exported to stackdriver/ GCM, it seems many of the `Int64ObservableGauge` that get registered via callbacks don't show up. At first I thought this was GCM exporter specific but looking at the prometheus exporter, Jay and I saw some of those metrics didn't show up on the `/metrics` endpoint either.

Digging into the call paths a bit, calling `RegisterCallback` we might get back a [noop impl](https://github.com/open-telemetry/opentelemetry-go/blob/4491b39db28770af8c1b2bec183cfaa426ef4b1f/metric.go#L39-L45):

```
// If no global GetMeterProvider has been registered, a No-op GetMeterProvider
// implementation is returned. When a global GetMeterProvider is registered for
// the first time, the returned GetMeterProvider, and all the Meters it has
// created or will create, are recreated automatically from the new
// GetMeterProvider.
```

This reads very race-y to me given that we set a meter provider when the manager `Start`s the exporter runnables but we `.GetMeterProvider` throughout the gatekeeper controllers and subsystems (audit, etc.). I think the key here would be to have the `MeterProvider` set before we attempt to use it throughout Gatekeeper. 

I played around locally and initialized the provider as part of a `init()` call and was able to see all metrics, including those registered via callabacks, show up consistently in gcm.

Reply:
Right now, the flow of the execution makes it so that `MeterProvider` is set before individual metrics are getting initialized. I made some changes for the metrics that were troublesome. I think it should be fixed now. Let's sync up on Tuesday to verify the same and confirm the change.

Reply:
Hm I still think there is some sort of race between `SetMeterProvider` and `GetMeterProvider`.  From my testing, it is clear to me that some meter, i.e. calls to `GetMeterProvider().Meter()` get the no-op implementation, hence why the callback style metrics don't show up.

Unfortunately, this seems to still be an issue as of https://github.com/open-policy-agent/gatekeeper/pull/3011/commits/e4a707f4be6722128c705a3c6d4a2d21228b0d99 for both `prom` and `gcm/ stackdriver`

I think want we really want is to have `SetMeterProvider` get called before we attempt to `GetMeterProvider` in any of the packages w stats_reporter. I'm not going to provide a strong recommendation but we could consider moving the exporters initialization in an `init()` func (and not have it be predicated by the manager to start them as runnables) but that would require some significant re-jigging. Also, I don't see a stated contract  the order of  nested packages' `init` execution on https://go.dev/ref/mem#init. Alternatively, we could  lazy register the metrics "later", after `setupControllers` is done & we run the `mgr,Start()` in main. Maybe let's discuss where we are at the next community meeting ?

Reply:
@acpana Refactored to initiate measure/metrics inside `newReporter()` instead of `init()`, Let's see if this problem still persist.